selfdrive/ui/qt/maps/map_instructions.cc

```diff
@@ -89,6 +89,10 @@ void MapInstructions::updateInstructions(cereal::NavInstruction::Reader instruct
     QString fn = "direction_" + type;
     if (!modifier.isEmpty()) {
       fn += "_" + modifier;
+      if (type == "roundabout" || type == "rotary"){
+        // Normalize and round to nearest 45deg
+        fn += "_" + QString::number(((instruction.getManeuverDegrees() + 382) % 360 / 45 * 45));
+      }
```

Comment:
does it make more sense to fix in navd instead?

Reply:
Hmm I would say it is simpler if its done here. I guess the issue is that the modification won't get captured in the logs. If that's important, then maybe its worth moving the math to navd.py.

Reply:
Shouldn't need to rename the icons as well, the maneuver direction should just be changed

Reply:
> I think it's worth fixing at the source (navd), those instructions are used in the model as well

Good point however would it not be better if the model had the unmodified data from the response? The rounding is only needed to display the limited number of icons that are available.

> Shouldn't need to rename the icons as well, the maneuver direction should just be changed

The maneuver direction is for the entry angle and does not represent which exit should be taken. Mapbox could specify slight_right roundabout 180 which would not display anything since there is no icon for it.

![image](https://github.com/commaai/openpilot/assets/83676301/af4ca078-1993-4dbd-bcf8-29f2421c8190)

if you use roundabout_exits=true in the request it will output a separate maneuver for the exit as well.  So technically you could show a different icon for every combination of entry/exit maneuvers and degrees.

> Let's fix in navd, a field only specific to one maneuver doesn't sound the cleanest

I agree it doesn't seem like great design but it keeps the code aligned to the mapbox spec and the current code style. The other way this can be done is if you map the degrees to a maneuver name which eliminates the need for the degree field at the cost of accuracy/clarity. This is how I thought of doing it initially but its a kludge imo that does pan out if you want more detailed icons/model data in the future.

I will make another PR that doesn't add a cereal field and does it in navd.py and you can choose what's better.

Reply:
If I remove the degrees field in cereal and overwrite the maneuver modifier i gotta do something like this.

in helpers.py
```
if field_valid(p, 'degrees'):
  if p['type'] == 'roundabout' or p['type'] == 'rotary':
    directions = [
      'straight',     # 0 or 360
      'slight_right', # 45
      'right',        # 90
      'tight_right',  # 135
      'straight',     # 180
      'slight_left',  # 225
      'left',         # 270
      'tight_left'    # 315
    ]
    deg = (p['degrees'] + 360) % 360
    i = (deg + 22) // 45 % 8
    instruction['maneuverModifier'] = directions[i]
```

I don't like this method but if you think the model is better suited to this method then I can submit the PR for it.

I was also thinking that you can use the degree field for other maneuvers by filling it with the bearing_before - bearing_after.

Reply:
So I just round this conversation but this is what I have been using to resolve this issue in navd helpers.py
![roundabout](https://github.com/commaai/openpilot/assets/98910897/7a8ec1cf-bac6-4d20-be7b-c0983653729d)
``` 
  if field_valid(p, 'degrees'):
    maneuverDegrees = int(p['degrees'])
    drivingSide = p['driving_side']
    modifier1 = ""
    modifier2 = ""
    if (0 <= maneuverDegrees <= 67) or (292 < maneuverDegrees <= 360):
      modifier1 = "sharp "
    elif (112 < maneuverDegrees <= 157) or (202 < maneuverDegrees <= 247):
      modifier1 = "slight "
    if 0 <= maneuverDegrees <= 157:
      modifier2 = drivingSide
    elif 157 < maneuverDegrees <= 202:
      modifier2 = "straight"
    elif 202 < maneuverDegrees <= 360:
      modifier2 = "left" if drivingSide == "right" else "right"
    instruction['maneuverModifier'] = modifier1 + modifier2

Reply:
> So I just found this conversation but this is what I have been using to resolve this issue in navd helpers.py I am not using any additional icons than stock and it seems to work pretty well. ![roundabout](https://private-user-images.githubusercontent.com/98910897/272143612-7a8ec1cf-bac6-4d20-be7b-c0983653729d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMyMzE1MDEsIm5iZiI6MTcwMzIzMTIwMSwicGF0aCI6Ii85ODkxMDg5Ny8yNzIxNDM2MTItN2E4ZWMxY2YtYmFjNi00ZDIwLWJlN2ItYzA5ODM2NTM3MjlkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIyVDA3NDY0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4MDE2YTgzZjU4NDA5YzNlN2QwMjQxN2E2ZmEzMDI3ZTNhZmJkZjM4ZTAzYjk1NjI5MzQzYTI1OGM1NzkzNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.32Q_AjEbb52TMWclgMNFuUK3w6r9zDfELMCd2z6XeRs)
> 
> ```
>   if field_valid(p, 'degrees'):
>     maneuverDegrees = int(p['degrees'])
>     drivingSide = p['driving_side']
>     modifier1 = ""
>     modifier2 = ""
>     if (0 <= maneuverDegrees <= 67) or (292 < maneuverDegrees <= 360):
>       modifier1 = "sharp "
>     elif (112 < maneuverDegrees <= 157) or (202 < maneuverDegrees <= 247):
>       modifier1 = "slight "
>     if 0 <= maneuverDegrees <= 157:
>       modifier2 = drivingSide
>     elif 157 < maneuverDegrees <= 202:
>       modifier2 = "straight"
>     elif 202 < maneuverDegrees <= 360:
>       modifier2 = "left" if drivingSide == "right" else "right"
>     instruction['maneuverModifier'] = modifier1 + modifier2
> ```

Doing it this way will change the model inputs.

Reply:
If you aren’t trying to change model inputs and solely going after navigation icons for roundabouts, I suggest moving to what Waze does and use the simple roundabout icon with exit number in the center. That can be extracted very easily from the map box api provided. It also lines up with the human instructions they provide as well. 
After driving on my version of the degrees fix, map box does not always supply the correct number of degrees for roundabouts. Exit number is always accurate though. 
Just a thought.

Reply:
Can you provide a route where mapbox doesn't supply the correct angle. If its an issue with mapbox then we can submit an edit. if its an issue with parsing then we need to fix it.

Can you provide an example roundabout where it provides the wrong degrees value.